limiter/queue_blocking.go

```diff
@@ -0,0 +1,313 @@
+package limiter
+
+import (
+	"container/list"
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/platinummonkey/go-concurrency-limits/core"
+)
+
+// QueueOrdering defines the pattern for ordering requests on Pool
+type QueueOrdering string
+
+// The available options
+const (
+	OrderingFIFO QueueOrdering = "fifo"
+	OrderingLIFO QueueOrdering = "lifo"
+)
+
+type queueElement struct {
+	ctx         context.Context
+	releaseChan chan<- core.Listener
+	next, prev  *queueElement
+	evicted     bool
+}
+
+func (e *queueElement) setListener(listener core.Listener) bool {
+	select {
+	case e.releaseChan <- listener:
+		close(e.releaseChan)
+		return true
+	default:
+		// timeout has expired
+		return false
+	}
+}
+
+func (q *queue) evictionFunc(e *list.Element) func() {
+	return func() {
+		q.mu.Lock()
+		defer q.mu.Unlock()
+
+		qe := e.Value.(*queueElement)
+		// Prevent multiple invocations from
+		// corrupting the queue state.
+		if qe.evicted {
+			return
+		}
+
+		qe.evicted = true
+		q.list.Remove(e)
+	}
+}
+
+type queue struct {
+	list     *list.List
+	ordering QueueOrdering
+	mu       sync.RWMutex
+}
+
+func (q *queue) len() uint64 {
+	q.mu.RLock()
+	defer q.mu.RUnlock()
+	return uint64(q.list.Len())
+}
+
+func (q *queue) push(ctx context.Context) (func(), <-chan core.Listener) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	releaseChan := make(chan core.Listener)
+
+	e := &queueElement{ctx: ctx, releaseChan: releaseChan}
+	listElement := q.list.PushFront(e)
+
+	return q.evictionFunc(listElement), releaseChan
+}
+
+func (q *queue) pop() *queueElement {
+	evict, ele := q.peek()
+	if evict != nil {
+		evict()
+	}
+
+	return ele
+}
+
+func (q *queue) peek() (func(), *queueElement) {
+	q.mu.RLock()
+	defer q.mu.RUnlock()
+	if q.list.Len() > 0 {
+
+		var element *list.Element
+		switch q.ordering {
+		case OrderingFIFO:
+			element = q.list.Back()
+		case OrderingLIFO:
+			element = q.list.Front()
+		}
```

Comment:
This is essentially the biggest difference between this impl and what exists in the lifo impl. It is also the only code path where we have any logic dependent on the queue order